src/Components/Components/src/Sections/SectionRegistry.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Components.Sections
+{
+    internal class SectionRegistry
+    {
+        private readonly Dictionary<string, List<ISectionContentSubscriber>> _subscribersByName = new();
+        private readonly Dictionary<string, List<ISectionContentProvider>> _providersByName = new();
+
+        public void AddProvider(string name, ISectionContentProvider provider)
+        {
+            if (!_providersByName.TryGetValue(name, out var providers))
+            {
+                providers = new();
+                _providersByName.Add(name, providers);
+            }
+
+            providers.Add(provider);
+        }
+
+        public void RemoveProvider(string name, ISectionContentProvider provider)
+        {
+            if (!_providersByName.TryGetValue(name, out var providers))
+            {
+                throw new InvalidOperationException($"There are no content providers with the name '{name}'.");
+            }
+
+            var index = providers.LastIndexOf(provider);
+
+            if (index < 0)
+            {
+                throw new InvalidOperationException($"The provider was not found in the providers list of name '{name}'.");
+            }
+
+            providers.RemoveAt(index);
+
+            if (index == providers.Count)
+            {
+                // We just removed the most recently added provider, meaning we need to change
+                // the current content to that of second most recently added provider.
+                var content = GetCurrentProviderContentOrDefault(providers);
+                NotifyContentChangedForAllSubscribers(name, content);
+            }
+        }
+
+        public void Subscribe(string name, ISectionContentSubscriber subscriber)
```

Comment:
I'll let @SteveSandersonMS chime in on this, but I'd prefer if the implementation stuck to one subscriber - it simplifies this implementation and doesn't create a feature that is currently unused (and untested).